### PR TITLE
JSG_STRUCT_TS_OVERRIDE_DYNAMIC by refactoring JSG_STRUCT

### PR DESCRIPTION
**Pull request by AdityaAtulTewari**
# Enable Dynamic TS overrides

## Goal:
Enable Dynamic typescript overrides that are based on compat flags
```cpp
JSG_STRUCT(.........);
JSG_STRUCT_TS_OVERRIDE_DYNAMIC(CompatibilityFlags::Reader flags) {
  if(flags.getCacheEnabled()) {
    JSG_TS_OVERRIDE(...);
   else {
    JSG_TS_OVERRIDE(...);
   }
}
```

## Mechanism:
We do this by refactoring JSG_STRUCT to be more extensible in line with `JSG_RESOURCE_TYPE`.
We create an internal function `registerMembersInternal` that enables no repeat code and we use `c++20` requires in order to select which implementation of `registerMembers` should be selected.

### Steps:
- [x] Refactor JSG_STRUCT with an internal function
- [x] Add JSG_STRUCT_TS_OVERRIDE_DYNAMIC(..) macro
- [x] Create another INTERNAL macro that conditionally includes that behavior
- [x] Test

### Test plan
- [x] Successful use in #2434 
- [ ] Add another test to `jsg-test.c++`

### CODE REVIEW

**Comment by jasnell** [comment]
I'm good with this approach but would like @kentonv to review and sign off.

**Comment by AdityaAtulTewari** [comment]
> Any chance of a test?

I want to add a test but how would it work? It is hard to automate a test to check whether or not types exist. I'm happy to do something if you can suggest how to test the overrides.

**Review by kentonv — COMMENTED** [review]
Is the motivation for this to have `RequestInitializerDict::cache`'s type change based on the newly-created compat flag?

Is that actually important? It seems like we can just define the type according to the standard either way -- but without the compat flag, it'll throw (at runtime) if you actually set it.

File: src/workerd/jsg/jsg.h [hunk 1]
```diff
@@ -623,6 +630,41 @@ namespace {
   struct HasStructTypeScriptDefine<T, decltype(T::_JSG_STRUCT_TS_DEFINE_DO_NOT_USE_DIRECTLY, 0)> : std::true_type { };
 }
 
+// This function is the first part of a JSG_STRUCT and defines the Kind, fields, and type wrapper
+#define _JSG_STRUCT_INTERNAL_KIND_AND_FIELDS_AND_TYPE_WRAPPER(...) \
```
[the remark this answers is not in the document]
Comment (AdityaAtulTewari) [reply src/workerd/jsg/jsg.h:634]: ah I should change that.

File: src/workerd/jsg/jsg.h [hunk 2]
```diff
@@ -649,26 +705,23 @@ namespace {
...
+    _JSG_STRUCT_REGISTER_MEMBERS_TS_ROOT(); \
+    _JSG_STRUCT_REGISTER_MEMBERS_TS_OVERRIDE(); \
+    _JSG_STRUCT_REGISTER_MEMBERS_TS_DEFINE(); \
   } \
-  template <typename TypeWrapper, typename Self> \
-  using JsgFieldWrappers = ::workerd::jsg::TypeTuple< \
-    JSG_FOR_EACH(JSG_STRUCT_FIELD, , __VA_ARGS__) \
-  >
+  template <typename Registry, typename Self> \
+  static void registerMembers(Registry& registry, jsg::GetConfiguration<Self> arg) \
```
[the remark this answers is not in the document]
Comment (AdityaAtulTewari) [reply src/workerd/jsg/jsg.h:718]: because we need the two argument version of the function.

**Comment by AdityaAtulTewari** [comment]
> Is the motivation for this to have `RequestInitializerDict::cache`'s type change based on the newly-created compat flag?
Yes in the generated typescript.

> Is that actually important? It seems like we can just define the type according to the standard either way -- but without the compat flag, it'll throw (at runtime) if you actually set it.
My motivation for this is to create total equality to the initial state so without the flag the users sees no difference. However @jasnell  may have other reasons for this.

File: src/workerd/jsg/jsg.h (hunk 2 above)
[the remark this answers is not in the document]
Comment (kentonv) [reply src/workerd/jsg/jsg.h:718]: But we didn't need a two-arg version before? What changed?

[the remark this answers is not in the document]
Comment (AdityaAtulTewari) [reply src/workerd/jsg/jsg.h:718]: we need it to pass the compatibility flag to the DYNAMIC_OVERRIDE.

[the remark this answers is not in the document]
Comment (AdityaAtulTewari) [reply src/workerd/jsg/jsg.h:718]: perhaps I should make the macro more explicit?

[the remark this answers is not in the document]
Comment (kentonv) [reply src/workerd/jsg/jsg.h:718]: Ah, at the time I looked, `JSG_STRUCT_REGISTER_MEMBERS_TS_DYNAMIC_OVERRIDE()` didn't take a parameter, so it looked like `arg` was unused and I couldn't figure out what the difference between the two methods. It's much easier to understand with the macro argument.

File: src/workerd/jsg/jsg.h [hunk 3]
```diff
@@ -649,26 +705,23 @@ namespace {
...
+      requires jsg::HasConfiguration<Self> { \
+    JSG_FOR_EACH(JSG_STRUCT_REGISTER_MEMBER, , __VA_ARGS__); \
+    JSG_STRUCT_REGISTER_MEMBERS_TS_ROOT(); \
```
Comment (kentonv) [inline src/workerd/jsg/jsg.h:721]: Instead of using macros for code sharing here, could the two versions of the method call a common implementation that is a template? These macros are pretty hard to follow -- and they pollute the global scope.

**Comment by kentonv** [comment]
> My motivation for this is to create total equality to the initial state so without the flag the users sees no difference. However @jasnell may have other reasons for this.

IMO it's not important to prevent any observable change to the typescript types. The important thing is that we don't break workers in production, but typescript changes don't have any effect on what's already in production.

File: src/workerd/jsg/jsg.h (hunk 3 above)
Comment (AdityaAtulTewari) [reply src/workerd/jsg/jsg.h:721]: yeah I think there is a way to do that, that suggestion makes sense.

Comment (AdityaAtulTewari) [reply src/workerd/jsg/jsg.h:721]: yeah I think there is a way to do that, that suggestion makes sense.

**Comment by jasnell** [comment]
> ... it's not important to prevent any observable change to the typescript types...

I disagree with this a bit. We've had a number of [issues](https://github.com/cloudflare/workerd/issues?q=sort%3Aupdated-desc+is%3Aissue+is%3Aclosed+types) pop up due to observable changes in the types and it is something that we at least should be careful of. I don't feel strongly about this specific change, I had mentioned only to make sure any possible issues were considered. I would defer to the frameworks team (@IgorMinar @dario-piotrowicz @jculvey @vicb ) to weigh in on what to do here.

**Comment by kentonv** [comment]
I maintain that it is not important to ensure that `RequestInitializerDict.cache` is absent from the type declarations when the compat flag is not enabled. Having this declaration appear in the types won't break anyone in itself, even if attempts to use it throw an exception. Sure, there's a remote chance it causes some minor confusion for someone writing new code, but probably that confusion exists already as people are probably looking at MDN docs, where this property has been defined and documented all along.

So I don't think we need to bend over backwards adding new features to JSG just to avoid this.

**Comment by penalosa** [comment]
> I maintain that it is not important to ensure that `RequestInitializerDict.cache` is absent from the type declarations when the compat flag is not enabled. Having this declaration appear in the types won't break anyone in itself, even if attempts to use it throw an exception. Sure, there's a remote chance it causes some minor confusion for someone writing new code, but probably that confusion exists already as people are probably looking at MDN docs, where this property has been defined and documented all along.

I don't really agree with this. The main idea behind all of our TypeScript generation work has been to make sure that the TS types match runtime behaviour as closely as possible (and the recent `wrangler types` improvements have gone further in that direction). If we _can_, it would be a much nicer DX to have `.cache` cause a type error with the wrong compat date, rather than relying on a runtime error. However, I'm not sure I really understand why this needs JSG changes—we already do this type of thing. For instance, the global navigator is [only added when the right compat flag is enabled](https://github.com/cloudflare/workerd/blob/main/src/workerd/api/global-scope.h#L604-L607):

```c++
if (flags.getGlobalNavigator()) {
  JSG_LAZY_INSTANCE_PROPERTY(navigator, getNavigator);
  JSG_NESTED_TYPE(Navigator);
}
```

**Comment by jasnell** [comment]
@penalosa 
> ... If we can, it would be a much nicer DX to have .cache cause a type error with the wrong compat date, rather than relying on a runtime error. However, I'm not sure I really understand why this needs JSG changes—we already do this type of thing. For instance, the global navigator is [only added when the right compat flag is enabled](https://github.com/cloudflare/workerd/blob/main/src/workerd/api/global-scope.h?rgh-link-date=2024-08-20T15%3A18%3A34Z#L604-L607):

the difference here is that `JSG_RESOURCE_TYPE` already implements the ability to vary the type output based on compat flags and `JSG_STRUCT` does not. We also have no way of conditionally handling the `cache` property on a `JSG_STRUCT` based on the compat flag settings. So the options right now are: (a) always expose the new types for `cache` regardless of whether the compat flag is set or (b) land the changes in this PR to make it respect the compat flag.

**Review by kentonv — APPROVED** [review]
The latest iteration of the code looks fine. Still question the necessity but I won't block it.

**Review by jasnell — APPROVED** [review]
